File: apps/android/app/src/main/java/chat/rogi/rogichat/core/messageactions/ActorBlocks.kt
package chat.rogi.rogichat.core.messageactions

import chat.rogi.rogichat.core.auth.StrictAuthJson
import java.time.Instant
import java.util.UUID
import kotlinx.serialization.json.*

/** Own blocks/unblock remains available after leaving; no active membership token is required. */
data class BlockScope(val environment: String, val accountId: String, val sessionEpoch: String, val roomId: String, val viewEpoch: String) {
    init { require(environment in setOf("qa", "prod")); listOf(accountId, sessionEpoch, roomId, viewEpoch).forEach(::actionId) }
    val partition get() = listOf(environment, accountId, roomId)
}
/** Current own-block GET projection only; never enrich from message/profile/history caches. */
data class BlockedActor(val actorId: String, val blockedAt: String, val displayName: String?) {
    val displayLabel: String get() = displayName ?: "이름을 확인할 수 없는 사용자"
}
data class BlockPage(val blocks: List<BlockedActor>, val next: String?)
enum class UnblockOutcome { UNKNOWN, ACKNOWLEDGED, REJECTED }
data class UnblockRecord(val id: String, val scope: BlockScope, val actorId: String,
    val outcome: UnblockOutcome, val observedBlocked: Boolean? = null)
interface BlockJournal { fun records(): List<UnblockRecord>; fun put(record: UnblockRecord) }
class BlockViewToken internal constructor(val scope: BlockScope, internal val generation: Long) {
    override fun equals(other: Any?) = other is BlockViewToken && scope == other.scope && generation == other.generation
    override fun hashCode() = 31 * scope.hashCode() + generation.hashCode()
}
class BlockPageToken internal constructor(val view: BlockViewToken, val after: String?)
class UnblockPermit internal constructor(val record: UnblockRecord, private val owner: ActorBlocksState, private val token: BlockViewToken) {
    private var used = false
    fun claim() = synchronized(owner) { check(!used && owner.admits(token) && owner.pending === this); used = true }
    fun request() = ActionRequest("DELETE", "rooms/${record.scope.roomId}/blocks/${record.actorId}", null, 200)
}
/** Return Reset on acknowledged change OR fresh current-state recovery, never infer operation success from a GET. */
data class BlockReset(val scope: BlockScope)
class ActorBlocksState(private val journal: BlockJournal, private val actionJournal: ActionJournal) {
    private var scope: BlockScope? = null
    private var generation = 0L
    private var pagePending: BlockPageToken? = null
    var pending: UnblockPermit? = null; private set
    var blocks: List<BlockedActor> = emptyList(); private set
    var complete = false; private set
    var next: String? = null; private set
    var failed = false; private set
    var lastOutcome: UnblockOutcome? = null; private set
    private var baseline: List<ActionRecord> = emptyList()
    @Synchronized fun select(value: BlockScope?) {
        generation++; scope = value; pending = null; pagePending = null; blocks = emptyList(); complete = false; next = null; failed = false; lastOutcome = null; baseline = emptyList()
    }
    @Synchronized fun capture() = scope?.let { BlockViewToken(it, generation) }
    internal fun admits(token: BlockViewToken) = token.scope == scope && token.generation == generation
    @Synchronized fun refresh(): BlockPageToken? {
        if (pending != null || scope == null) return null
        baseline = currentActorRecords()
        generation++; blocks = emptyList(); complete = false; next = null; failed = false
        return BlockPageToken(capture()!!, null).also { pagePending = it }
    }
    @Synchronized fun more(): BlockPageToken? {
        if (pending != null || pagePending != null || complete || next == null) return null
        return BlockPageToken(capture() ?: return null, next).also { pagePending = it; failed = false }
    }
    @Synchronized fun fail(token: BlockPageToken) {
        if (pagePending === token && admits(token.view)) { pagePending = null; failed = true }
    }
    @Synchronized fun accept(token: BlockPageToken, page: BlockPage): BlockReset? {
        if (pagePending !== token || !admits(token.view)) return null
        require(page.blocks.size <= 50 && page.blocks.map { it.actorId }.distinct().size == page.blocks.size)
        page.blocks.forEach { actionId(it.actorId); Instant.parse(it.blockedAt) }
        require(page.blocks.zipWithNext().all { (a,b) -> a.actorId < b.actorId })
        require(token.after == null || page.blocks.all { it.actorId > token.after })
        require(page.next == null || page.next == page.blocks.lastOrNull()?.actorId)
        blocks = if (token.after == null) page.blocks else blocks + page.blocks
        next = page.next; complete = next == null; pagePending = null; failed = false
        if (currentActorRecords() != baseline) {
            generation++; blocks = emptyList(); complete = false; next = null; failed = true
            return null // A block intent/receipt changed while this GET was in flight.
        }
        if (!complete) return null // Prefix omission never proves unblocked.
        val current = token.view.scope; val actors = blocks.map { it.actorId }.toSet()
        journal.records().filter { it.scope.partition == current.partition && it.outcome == UnblockOutcome.UNKNOWN }
            .forEach { journal.put(it.copy(observedBlocked = it.actorId in actors)) }
        actionJournal.records().filter { it.action == MessageAction.BLOCK_ACTOR &&
            listOf(it.selection.scope.environment, it.selection.scope.accountId, it.selection.scope.roomId) == current.partition }
            .forEach { actionJournal.put(it.copy(observedBlocked = it.selection.visibleActorId in actors)) }
        return BlockReset(current)
    }
    private fun currentActorRecords(): List<ActionRecord> = actionJournal.records().filter { it.action == MessageAction.BLOCK_ACTOR &&
        listOf(it.selection.scope.environment, it.selection.scope.accountId, it.selection.scope.roomId) == scope?.partition }.sortedBy { it.id }
    @Synchronized fun unknownActors(): Set<String> {
        val current = scope ?: return emptySet()
        return journal.records().filter { it.scope.partition == current.partition && it.outcome == UnblockOutcome.UNKNOWN }.map { it.actorId }.toSet() +
            actionJournal.records().filter { it.action == MessageAction.BLOCK_ACTOR && it.phase == ActionPhase.UNKNOWN &&
                listOf(it.selection.scope.environment, it.selection.scope.accountId, it.selection.scope.roomId) == current.partition }.mapNotNull { it.selection.visibleActorId }
    }
    /** Explicit user intent after a COMPLETE fresh list. Previous UNKNOWN remains UNKNOWN in history. */
    @Synchronized fun unblock(token: BlockViewToken, actorId: String): UnblockPermit {
        check(admits(token) && complete && pending == null && blocks.any { it.actorId == actorId })
        val record = UnblockRecord(UUID.randomUUID().toString(), token.scope, actorId, UnblockOutcome.UNKNOWN)
        journal.put(record); lastOutcome = UnblockOutcome.UNKNOWN; generation++; complete = false; pagePending = null
        return UnblockPermit(record, this, capture()!!).also { pending = it }
    }
    @Synchronized fun finish(permit: UnblockPermit, outcome: UnblockOutcome): BlockReset? {
        val original = journal.records().singleOrNull { it.id == permit.record.id } ?: return null
        if (original.outcome != UnblockOutcome.UNKNOWN) return null
        journal.put(original.copy(outcome = outcome))
        if (pending !== permit || scope != original.scope) return null
        pending = null; lastOutcome = outcome
        if (outcome != UnblockOutcome.ACKNOWLEDGED) return null
        blocks = blocks.filterNot { it.actorId == original.actorId }
        return BlockReset(original.scope)
    }
}
object ActorBlocksWire {
    fun list(token: BlockPageToken): ActionRequest {
        token.after?.let(::actionId)
        return ActionRequest("GET", "rooms/${token.view.scope.roomId}/blocks", null, 200, token.after?.let { mapOf("after" to it) }.orEmpty())
    }
    fun page(body: String): BlockPage {
        val root = StrictAuthJson.objectValue(body); require(root.keys == setOf("blocks", "next"))
        val rows = root.getValue("blocks").jsonArray.map { raw ->
            val row = raw.jsonObject; require(row.keys == setOf("actorId", "blockedAt", "displayName"))
            BlockedActor(row.text("actorId").also(::actionId), row.text("blockedAt").also { Instant.parse(it) },
                if (row.getValue("displayName") == JsonNull) null else row.text("displayName"))
        }
        require(rows.size <= 50)
        return BlockPage(rows, if (root.getValue("next") == JsonNull) null else root.text("next").also(::actionId))
    }
    fun result(permit: UnblockPermit, status: Int, body: String): UnblockOutcome = try {
        val root = StrictAuthJson.objectValue(body)
        if (status == 200) {
            require(root.keys == setOf("actorId", "blocked", "resetRequired") && root.text("actorId") == permit.record.actorId)
            val blocked = root.getValue("blocked").jsonPrimitive; val reset = root.getValue("resetRequired").jsonPrimitive
            require(!blocked.isString && blocked.booleanOrNull == false && !reset.isString && reset.booleanOrNull == true)
            UnblockOutcome.ACKNOWLEDGED
        } else if (MessageActionWire.result(MessageAction.BLOCK_ACTOR, status, body) == ActionResult.Rejected) UnblockOutcome.REJECTED else UnblockOutcome.UNKNOWN
    } catch (_: Exception) { UnblockOutcome.UNKNOWN }
    private fun JsonObject.text(key: String) = getValue(key).jsonPrimitive.let { require(it.isString); it.content }
}

File: apps/ios/Sources/Core/MessageActions/ActorBlocks.swift
import Foundation
import CoreFoundation

/// Own block management remains available after leave; no active membership scope is needed.
struct BlockScope: Codable, Equatable, Sendable {
    let environment: String; let accountId: String; let sessionEpoch: String; let roomId: String; let viewEpoch: String
    var valid: Bool { ["qa", "prod"].contains(environment) && [accountId, sessionEpoch, roomId, viewEpoch].allSatisfy(actionID) }
    var partition: [String] { [environment, accountId, roomId] }
}
struct BlockedActor: Equatable, Sendable { let actorId: String; let blockedAt: String }
struct BlockPage: Equatable, Sendable { let blocks: [BlockedActor]; let next: String? }
enum UnblockOutcome: String, Codable, Sendable { case unknown, acknowledged, rejected }
struct UnblockRecord: Codable, Equatable, Sendable {
    let id: String; let scope: BlockScope; let actorId: String
    var outcome: UnblockOutcome; var observedBlocked: Bool?
}
@MainActor protocol BlockJournal { func records() throws -> [UnblockRecord]; func put(_ record: UnblockRecord) throws }
struct BlockViewToken: Equatable, Sendable { let scope: BlockScope; fileprivate let generation: UInt64 }
final class BlockPageToken: Sendable {
    let view: BlockViewToken; let after: String?
    fileprivate init(view: BlockViewToken, after: String?) { self.view = view; self.after = after }
}
final class UnblockPermit: @unchecked Sendable {
    let record: UnblockRecord
    private let lock = NSLock(); private var used = false; private var revoked = false
    fileprivate init(_ record: UnblockRecord) { self.record = record }
    func revoke() { lock.withLock { revoked = true } }
    func claim() throws { try lock.withLock { guard !used, !revoked else { throw MessageActionError.stale }; used = true } }
    func request() -> ActionRequest { ActionRequest(method: "DELETE", path: "rooms/\(record.scope.roomId)/blocks/\(record.actorId)", body: nil, successStatus: 200) }
}
struct BlockReset: Equatable, Sendable { let scope: BlockScope }
@MainActor final class ActorBlocksState {
    private let journal: any BlockJournal; private let actionJournal: any ActionJournal
    private var scope: BlockScope?; private var generation: UInt64 = 0
    private var pagePending: BlockPageToken?
    private(set) var pending: UnblockPermit?
    private(set) var blocks: [BlockedActor] = []
    private(set) var complete = false
    private(set) var next: String?
    private(set) var failed = false
    private(set) var lastOutcome: UnblockOutcome?
    private var baseline: [ActionRecord] = []
    init(journal: any BlockJournal, actionJournal: any ActionJournal) { self.journal = journal; self.actionJournal = actionJournal }
    func select(_ value: BlockScope?) throws {
        generation &+= 1; pending?.revoke(); pending = nil; pagePending = nil; scope = nil; blocks = []; complete = false; next = nil; failed = false; lastOutcome = nil; baseline = []
        if let value, !value.valid { throw MessageActionError.invalidResponse }; scope = value
    }
    func capture() -> BlockViewToken? { scope.map { BlockViewToken(scope: $0, generation: generation) } }
    private func admits(_ token: BlockViewToken) -> Bool { token.scope == scope && token.generation == generation }
    func refresh() throws -> BlockPageToken? {
        guard pending == nil, scope != nil else { return nil }
        baseline = try currentActorRecords()
        generation &+= 1; blocks = []; complete = false; next = nil; failed = false
        let token = BlockPageToken(view: capture()!, after: nil); pagePending = token; return token
    }
    func more() -> BlockPageToken? {
        guard pending == nil, pagePending == nil, !complete, let next, let view = capture() else { return nil }
        let token = BlockPageToken(view: view, after: next); pagePending = token; failed = false; return token
    }
    func fail(_ token: BlockPageToken) {
        if pagePending === token, admits(token.view) { pagePending = nil; failed = true }
    }
    func accept(_ token: BlockPageToken, page: BlockPage) throws -> BlockReset? {
        guard pagePending === token, admits(token.view) else { return nil }
        guard page.blocks.count <= 50, Set(page.blocks.map(\.actorId)).count == page.blocks.count,
              page.blocks.allSatisfy({ actionID($0.actorId) && ActorBlocksWire.validDate($0.blockedAt) }),
              zip(page.blocks, page.blocks.dropFirst()).allSatisfy({ $0.actorId < $1.actorId }),
              token.after.map({ after in page.blocks.allSatisfy { $0.actorId > after } }) ?? true,
              page.next == nil || page.next == page.blocks.last?.actorId else { throw MessageActionError.invalidResponse }
        blocks = token.after == nil ? page.blocks : blocks + page.blocks
        next = page.next; complete = next == nil; pagePending = nil; failed = false
        guard try currentActorRecords() == baseline else {
            generation &+= 1; blocks = []; complete = false; next = nil; failed = true
            return nil // A concurrent block intent/receipt invalidates this read observation.
        }
        guard complete else { return nil } // Prefix omission never proves unblocked.
        let current = token.view.scope; let actors = Set(blocks.map(\.actorId))
        for var record in try journal.records() where record.scope.partition == current.partition && record.outcome == .unknown {
            record.observedBlocked = actors.contains(record.actorId); try journal.put(record)
        }
        for var record in try actionJournal.records() where record.action == .blockActor &&
            [record.selection.scope.environment, record.selection.scope.accountId, record.selection.scope.roomId] == current.partition {
            record.observedBlocked = record.selection.visibleActorId.map(actors.contains) ?? false; try actionJournal.put(record)
        }
        return BlockReset(scope: current)
    }
    private func currentActorRecords() throws -> [ActionRecord] {
        try actionJournal.records().filter { $0.action == .blockActor &&
            [$0.selection.scope.environment, $0.selection.scope.accountId, $0.selection.scope.roomId] == scope?.partition }.sorted { $0.id < $1.id }
    }
    func unknownActors() throws -> Set<String> {
        guard let scope else { return [] }
        let unblocks = try journal.records().filter { $0.scope.partition == scope.partition && $0.outcome == .unknown }.map(\.actorId)
        let actions = try actionJournal.records().filter { $0.action == .blockActor && $0.phase == .unknown &&
            [$0.selection.scope.environment, $0.selection.scope.accountId, $0.selection.scope.roomId] == scope.partition }.compactMap { $0.selection.visibleActorId }
        return Set(unblocks + actions)
    }
    /// Explicit user intent after a COMPLETE fresh list, never automatic replay of UNKNOWN.
    func unblock(_ token: BlockViewToken, actorId: String) throws -> UnblockPermit {
        guard admits(token), complete, pending == nil, blocks.contains(where: { $0.actorId == actorId }) else { throw MessageActionError.stale }
        let record = UnblockRecord(id: UUID().uuidString.lowercased(), scope: token.scope, actorId: actorId, outcome: .unknown)
        try journal.put(record); lastOutcome = .unknown; generation &+= 1; complete = false; pagePending = nil
        let permit = UnblockPermit(record); pending = permit; return permit
    }
    func finish(_ permit: UnblockPermit, outcome: UnblockOutcome) throws -> BlockReset? {
        guard var original = try journal.records().first(where: { $0.id == permit.record.id }), original.outcome == .unknown else { return nil }
        original.outcome = outcome; try journal.put(original)
        guard pending === permit, scope == original.scope else { return nil }
        pending = nil; lastOutcome = outcome
        guard outcome == .acknowledged else { return nil }
        blocks.removeAll { $0.actorId == original.actorId }; return BlockReset(scope: original.scope)
    }
}
enum ActorBlocksWire {
    static func list(_ token: BlockPageToken) throws -> ActionRequest {
        if let after = token.after, !actionID(after) { throw MessageActionError.invalidResponse }
        return ActionRequest(method: "GET", path: "rooms/\(token.view.scope.roomId)/blocks", body: nil, successStatus: 200, query: token.after.map { ["after": $0] } ?? [:])
    }
    static func validDate(_ value: String) -> Bool {
        let formatter = ISO8601DateFormatter(); formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return formatter.date(from: value) != nil || ISO8601DateFormatter().date(from: value) != nil
    }
    static func page(_ data: Data) throws -> BlockPage {
        let root = try ActionJSON.object(data)
        guard Set(root.keys) == ["blocks", "next"], let rows = root["blocks"] as? [[String: Any]], rows.count <= 50 else { throw MessageActionError.invalidResponse }
        let blocks = try rows.map { row -> BlockedActor in
            guard Set(row.keys) == ["actorId", "blockedAt"], let id = row["actorId"] as? String, actionID(id),
                  let date = row["blockedAt"] as? String, validDate(date) else { throw MessageActionError.invalidResponse }
            return BlockedActor(actorId: id, blockedAt: date)
        }
        let next = root["next"] as? String
        guard root["next"] is NSNull || next.map(actionID) == true else { throw MessageActionError.invalidResponse }
        return BlockPage(blocks: blocks, next: next)
    }
    static func result(_ permit: UnblockPermit, status: Int, data: Data) -> UnblockOutcome {
        do {
            let root = try ActionJSON.object(data)
            if status == 200 {
                guard Set(root.keys) == ["actorId", "blocked", "resetRequired"], root["actorId"] as? String == permit.record.actorId,
                      let blocked = root["blocked"] as? NSNumber, CFGetTypeID(blocked) == CFBooleanGetTypeID(), !blocked.boolValue,
                      let reset = root["resetRequired"] as? NSNumber, CFGetTypeID(reset) == CFBooleanGetTypeID(), reset.boolValue else { return .unknown }
                return .acknowledged
            }
            return MessageActionWire.result(.blockActor, status: status, data: data) == .rejected ? .rejected : .unknown
        } catch { return .unknown }
    }
}
